libbpf-tools/offcputime.bpf.c

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2021 Wenbo Zhang
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "offcputime.h"
+
+#define PF_KTHREAD		0x00200000	/* I am a kernel thread */
+#define MAX_ENTRIES		10240
+
+const volatile bool kernel_threads_only = false;
+const volatile bool user_threads_only = false;
+const volatile __u64 max_block_ns = -1;
+const volatile __u64 min_block_ns = 1;
+const volatile pid_t targ_tgid = -1;
+const volatile pid_t targ_pid = -1;
+const volatile long state = -1;
+
+struct internal_key {
+	u64 start_ts;
+	struct key_t key;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, struct internal_key);
+	__uint(max_entries, MAX_ENTRIES);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(key_size, sizeof(u32));
+} stackmap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, struct key_t);
+	__type(value, u64);
+	__uint(max_entries, MAX_ENTRIES);
+} info SEC(".maps");
+
+static bool allow_record(struct task_struct *t)
+{
+	if (targ_tgid != -1 && targ_tgid != t->tgid)
+		return false;
+	if (targ_pid != -1 && targ_pid != t->pid)
+		return false;
+	if (user_threads_only && t->flags & PF_KTHREAD)
+		return false;
+	else if (kernel_threads_only && !(t->flags & PF_KTHREAD))
+		return false;
+	if (state != -1 && t->state != state)
+		return false;
+	return true;
+}
+
+SEC("tp_btf/sched_switch")
+int BPF_PROG(sched_switch, bool preempt, struct task_struct *prev,
+	     struct task_struct *next)
+{
+	struct internal_key *i_keyp, i_key;
+	s64 *deltap, delta;
+	u32 pid;
+
+	if (allow_record(prev)) {
+		pid = prev->pid;
+		i_key.start_ts = bpf_ktime_get_ns();
+		i_key.key.pid = pid;
+		i_key.key.tgid = prev->tgid;
+		i_key.key.user_stack_id = -1; /* will support it later */
+		i_key.key.kern_stack_id = bpf_get_stackid(ctx, &stackmap, 0);
+		bpf_probe_read_str(&i_key.key.comm, sizeof(prev->comm),
+				   prev->comm);
+		bpf_map_update_elem(&start, &pid, &i_key, 0);
```

Comment:
From

> If you are using recently added BTF_PROG_TYPE_TRACING BPF programs, then you have a smartness of BPF verifier on your side, which now understands and tracks BTF types natively and allows you to follow pointers and read kernel memory directly (and safely), avoiding bpf_probe_read() calls, so you don’t need compiler rewriting magic to get the same nice and familiar syntax.
> 
> Libbpf + BPF_PROG_TYPE_TRACING way:
> 
> pid_t pid = task->pid;

I tried to use `&prev->pid ` as the map key, but got BPF verifier's complain:


```
106: (85) call bpf_map_update_elem#2
R2 type=ptr_ expected=fp, pkt, pkt_meta, map_value
```

Is this as expected?